### PR TITLE
Move Constraints

### LINKED ISSUE #18: Implement advanced movement constraints
I envision to be roughly 4 phases of development as far as chess movement constraints go. I present them here:
- [x] Footprint of each piece. Generic rules that are characteristic of each piece type.
- [x] Board-sensitive generic constraints like:
    - [x] Movement blocked by other pieces
    - [x] Capturing possibilities for a piece
- [x] Special rules:
    - [x] A pawn's
        - [x] initial (optional) 2-step
        - [x] capturing logic
        - [x] blocked movement logic
        - [x] promotion logic
        - [x] en passant
    - [x] A king's castling
- [x] Advanced generic rules:
    - [x] King checking and checkmate
        - [x] King: movement constraints
        - [x] Defenders: either block a check or move out of the way (all other moves crossed-out)
        - [x] Pins: movements that put your king in check must be crossed out
        - [x] Skewers: if check cannot be blocked, move out of the way

This is more of a meta-issue, but I wanted everything move-related in one place. I might be forgetting something, so this will be updated.

A few words about generic constraints. All pieces are blocked by friendlies. Foes block too, but they are capturable, unlike friendlies. All pieces but the pawn, that is. The pawn may be blocked by a foe standing right in front of it, because pawns may move straight up but capture up-diagonally. The King may also be indirectly blocked by a foe, if said foe is defended (so capturing said foe will put the king in check, and you can't do that).

### PR DESCRIPTION
**Pull request by TrentHawkins**
Start working on advanced constraints like blocking and capturing.

### CODE REVIEW

**Comment by prototact** [comment]
I think the important part is the condition function. It makes sense to define a `condition: Callable[[Square], bool]` together with some data. Since `legal_moves` generates all possible moves from the selected square, it makes sense to check for
- board edge
- stops at enemy pieces, including the captured square
- stops at friendly pieces, excluding the occupied square.

```python
def list_moves(board: Board, selected_square: Square) -> list[tuple[Square, Piece]]:
    piece = board[selected_square]
    def condition(square: Square) -> bool:
        other_piece = None
        try:
            other_piece = board[square]
        except IndexError:
            return False
        return other_piece is None or other_piece.is_black != piece.is_black
    target_squares = list()
    for square in piece.legal_moves(selected_square, condition):
        other_piece = board[square]
        if other_piece is None or other_piece.is_black != piece.is_black:
            target_squares.append((square, other_piece))
     return target_squares 
```
This would require to modify the `legal_moves` method because, for `Range Pieces`, the pieces would keep going even if capturing an enemy piece.

**Comment by TrentHawkins** [comment]
> I think the important part is the condition function. It makes sense to define a `condition: Callable[[Square], bool]` together with some data. Since `legal_moves` generates all possible moves from the selected square, it makes sense to check for

> This would require to modify the `legal_moves` method because, for `Range Pieces`, the pieces would keep going even if capturing an enemy piece.

Interesting, wit some adaptation this seems to be a working function. Just some minor tweeks:

```python
def list_moves(board: Board, selected_square: Square) -> set[tuple[Square, Piece]]:
    piece = board[selected_square]
    def condition(square: Square) -> bool:
        other_piece = None
        try:
            other_piece = board[square]
        except IndexError:
            return False
        return other_piece is None or other_piece.is_black != piece.is_black
    target_squares = set()
    for square in piece.legal_moves(selected_square, condition):
        other_piece = board[square]
        if other_piece is None or other_piece.color != piece.color:
            target_squares.add((square, other_piece))
     return target_squares
```

Thos tuples are expected to be unique and tuples are immutable, so set is in compliance with our general logic of collecting legal moves. Otherwise, lets keep this around.

File: src/square.py [hunk 1]
```diff
@@ -78,20 +78,38 @@ class Square(Vector):
...
+        """
+        try:
+            if isinstance(square, str):
+                assert Square.notation_range.match(square)
+            if isinstance(square, Vector):
+                assert 0 <= square.rank < 8 and 0 <= square.file < 8
+            return super(Square, cls).__new__(cls)
+        except AssertionError:
+            return None
+
```
Comment (TrentHawkins) [inline src/square.py]: `Square` objects are now always self-constrained within proper chessboard boundaries, discarding a trivial high level check for legal moves.

File: src/square.py [hunk 2]
```diff
@@ -112,14 +130,3 @@ def __repr__(self):
...
-    def is_in_board(self) -> bool:
-        """Agnostic conditions that apply to all squares.
-
-        In particular this contains boundary checks, since they are decipherable from square notation basically.
-        NOTE: This could go to `Board` instead too, I'm just trying to semantically allocate conditioning across objects.
-
-        Returns:
-            If square respects conditions.
-        """
-        return 0 <= self.rank < 8 and 0 <= self.file < 8
```
Comment (TrentHawkins) [inline src/square.py:125]: I rendered this method obsolete my making `Square` objects self-constrained. See [`__new__` method](https://github.com/TrentHawkins/Chess/pull/17/commits/7887e1699eef7117f73a6d58f8895cc2bea97208#r960090071).

File: src/piece.py [hunk 3]
```diff
@@ -167,8 +167,9 @@ def legal_moves(self, square: Square, condition: Callable[[Square], bool]) -> se
         super().legal_moves.__doc__
         squares = set()
         for step in self.steps:
-            if condition(square + step * self.color.value):
-                squares.add(square + step * self.color.value)
+            square_step = square + step * self.color.value
+            if square_step is not None and condition(square_step):
+                squares.add(square_step)
         return squares
```
Comment (TrentHawkins) [inline src/piece.py]: This is still messy, just mashing up a `Square`s self-constraint and `condition` closure in the conditional clauses. Just wanted to make sure it works, these parts need refactoring.

File: src/piece.py [hunk 4]
```diff
@@ -185,8 +186,9 @@ def legal_moves(self, square: Square, condition: Callable[[Square], bool]) -> se
         super().legal_moves.__doc__
         squares = set()
         for step in self.steps:
-            if condition(square + step):
-                squares.add(square + step)
+            square_step = square + step
+            if square_step is not None and condition(square_step):
+                squares.add(square_step)
         return squares
```
Comment (TrentHawkins) [inline src/piece.py]: See [this](https://github.com/TrentHawkins/Chess/pull/17/commits/7887e1699eef7117f73a6d58f8895cc2bea97208#r960091146).

File: src/piece.py [hunk 5]
```diff
@@ -263,11 +265,11 @@ class Range(Piece):
...
-            advanced_square = square + move  # Start looking at advanced squares
-            while condition(advanced_square):  # While we don't hit something.
-                squares.add(advanced_square)  # Add the damn square to legal destination squares.
-                advanced_square += move  # Advance to the next square in line.
+        for step in self.steps:  # For each direction.
+            square_step = square + step  # Start looking at advanced squares
+            while square_step is not None and condition(square_step):  # While we don't hit something.
+                squares.add(square_step)  # Add the damn square to legal destination squares.
+                square_step += step  # Advance to the next square in line.
         return squares
```
Comment (TrentHawkins) [inline src/piece.py]: See [this](https://github.com/TrentHawkins/Chess/pull/17/commits/7887e1699eef7117f73a6d58f8895cc2bea97208#r960091146).

**Comment by TrentHawkins** [comment]
A few remarks about the latest commit. I tried to stuff the simplest constraint for `legal_moves` (generated squares within bounds of a board) by forcing squares to return a None object when invalid coordinates are given for their creation. This in practice provides a natural stopping condition when operating on squares with vectors.

I believe there is a better way to do this, but I'm just putting this out there for you to review. To be frank, I am trying to stretch by python brain cells once in a while, and this `__new__` puzzle was a good exercise.

**Comment by TrentHawkins** [comment]
> I think the important part is the condition function. It makes sense to define a `condition: Callable[[Square], bool]` together with some data. Since `legal_moves` generates all possible moves from the selected square, it makes sense to check for
> 
> * board edge
> * stops at enemy pieces, including the captured square
> * stops at friendly pieces, excluding the occupied square.
> 
> ```python
> def list_moves(board: Board, selected_square: Square) -> list[tuple[Square, Piece]]:
>     piece = board[selected_square]
>     def condition(square: Square) -> bool:
>         other_piece = None
>         try:
>             other_piece = board[square]
>         except IndexError:
>             return False
>         return other_piece is None or other_piece.is_black != piece.is_black
>     target_squares = list()
>     for square in piece.legal_moves(selected_square, condition):
>         other_piece = board[square]
>         if other_piece is None or other_piece.is_black != piece.is_black:
>             target_squares.append((square, other_piece))
>      return target_squares 
> ```
> 
> This would require to modify the `legal_moves` method because, for `Range Pieces`, the pieces would keep going even if capturing an enemy piece.

Oh wait... I see now! Basically taking piece legal moves to generate a player's total legal moves, and performing the trimming at that level. Nice!

**Comment by TrentHawkins** [comment]
Since comparing weather two pieces are of the same subclass type and color is obsolete, I have overridden `Piece` `==` and `!=` to designate color difference. This operation could be left over for the `value` attribute along with inequalities even, but value is never to be compared, just stored and added (added I said eh?), so for now I make this notation useful for both `Piece` and `Board`.

See tests for an example:
```python
    def test_board_conditions(self):
        """Test the various logical conditions imposed at board level."""
        from src.board import Board
        board = Board()
        source = board["e2"]

        assert source is not None

        assert source != board["e2"] is False
        assert source != board["e4"] is False
        assert source != board["e7"] is True

        assert source == board["e2"] is True
        assert source == board["e4"] is False
        assert source == board["e7"] is False

```

File: src/square.py [hunk 6]
```diff
@@ -78,20 +78,38 @@ class Square(Vector):
...
+            `None` if square specification is illegal.
+        """
+        try:
+            if isinstance(square, str):
+                assert Square.notation_range.match(square)
+            if isinstance(square, Vector):
+                assert 0 <= square.rank < 8 and 0 <= square.file < 8
+            return super(Square, cls).__new__(cls)
+        except AssertionError:
+            return None
```
Comment (prototact) [inline src/square.py]: Catching assertion error to return `None` based on some truth value testing is kind of iffy. Α more direct way is
```python
        valid = False
        if isinstance(square, str):
            valid = Square.notation_range.match(square)
        elif isinstance(square, Vector):
            valid = square.rank in range(8) and square.file in range(8)
        if valid:
            return super().__new__(cls)
        return None
```

Comment (TrentHawkins) [reply src/square.py]: Agreed, this will be replaced. Commentary: `super` in class methods has to be explicitly be called with arguments, I think. I'll try again without arguments and see what happens.

**Comment by TrentHawkins** [comment]
Before I proceed with letting board handle conditioning, I will focus on how to compactify code logic and more specifically, get rid of all the `None` checks.

**Comment by prototact** [comment]
> Before I proceed with letting board handle conditioning, I will focus on how to compactify code logic and more specifically, get rid of all the `None` checks.

I think it all begins with the definition of `Square`, where a new `square` can be `None`. So `Square` is actually `Optional[Square]`.
In my opinion, defining `Square` alone is enough. The case where there is no `square` on the `board`, should be handled in the board class. In particular, 
- `square` is used to retrieve pieces from the board
- to change the position of a `piece` on the `board`.
 
Both of these use the `__getitem__` function of the board. The question is then
- what to do when there is no piece on the board (which returns `None`)
- what to do when the square is outside the board (an `IndexError` for the `list[list[Optional[Piece]]]` board)

This depends on the caller function. So which are the `caller` functions?
- a `move` function, that mutates the board. It takes a source and target `Square` as input.
- a `piece legal moves` function, that will determine whether the piece can do the `move`.
- a `capture_or_blocked` function, that somehow decides what to do if there is a `piece` on the `square`

For the first case, since the source and target squares are first checked by the `piece legal moves` and `capture_or_blocked` functions, I would assume the input is always good.

For the second case, there can be a check here that both source and target are valid squares. Otherwise, the source square is read from user input, so you could add the check there and only check for the target squares in the `piece legal moves` function.

Finally, the capture function will probably be called within the `piece legal moves` function when it finds a piece. Again, we can assume the input will always be valid (why check capture if there is no piece on the square, or the square does not exist?)

One other way to do this is to define a maybe function 
```python
from typing import TypeVar, Optional
from collections.abc import Callable


A = TypeVar("A")
B = TypeVar("B")


def maybe(a: Optional[A], f: Callable[[A], B], default: B) -> B:
    if a is None:
       return default
    return f(a)
```

An example would be `A = Square`, `f = board.__getitem__`, `B = Optional[Piece]`, `default = None`. However, this essentially changes the `a is None` checks with awkward `maybe` calls. It is better to sanitize your input and output when necessary, and write the "intra-domain" code with the assumption the input has only valid values.

In conclusion, I suggest:
- checking at user input whether the selected source square is valid, by catching an `IndexError`.
- checking in the `legal_moves` function, whether a target square is valid and whether a piece exists on the square. Again, the target square will fail with an `IndexError`.

Since we check in more than one place for the validity of a square, I would suggest to write a method for it on the board.
```python
def _check_square(self, square: Square) -> bool:
    valid = True
    try:
        self._board[square]
    except IndexError:
        valid = False
    return valid
```

**Comment by TrentHawkins** [comment]
@prototact I too prefer to sanitize input than to use a universal `None`-handler. I will follow your advice and try to trim the algebra as much as possible.

**Comment by TrentHawkins** [comment]
I went ahead with implementing movement blocking and capturing logic. Sanitizing `None` control is still outstanding, `None` sanity checks are all over the place. I should try to pull them down at a fundamental level and eliminate some cases, where wrong input should not be given and thus let an exception be thrown.

**Comment by TrentHawkins** [comment]
No success with `None` reduction. I only eliminated one late `None` check by incorporating it naturally into `Board.__getitem__` where it belongs. I want to “pull down” as many of these `None` checks as possible. Like the famous Python adage says:
> You (should) only have to say something once.

In other news, I have successfully drafted a method that uses two squares to form a move. I wish to refactor it to take a piece on the board and a target square.

Note that `Board.moves` keys all possible moves with the source square and not the source piece because Pieces are not immutable. But I think that in my current implementation I can safely throw away the `Piece.square` attribute and make them mutable again, with the only sacrifice being that `Piece.moves` methods will now assume a square argument. If that happens, I can key `Board.moves` with the board pieces instead (which is natural), which in turn will make `Board.move` naturally check amongst a player's pieces as well. Alternatively, check the commit coming right after for a shortcut that I might like.

**Comment by TrentHawkins** [comment]
And here it is. It is ugly, full of `None` checks, I will trim them down now that I have a basic working moving method. (No color check is added, that's the last thing to implement)

**Comment by TrentHawkins** [comment]
Hmmm, even if I remove the `Piece.square` attribute and make `Piece` objects hashable, their default hash will be by type. I tried unsafe hashing, but it failed, pieces remained identical by type even with the default `eq` defined by the `dataclass` decorator, which would normally compare `Piece` objects by all their attributes. I believe unsafe hashing looks and compares only the immutable components of the corresponding dataclass. Which means a manual `__hash__` may be required. This is not unreasonable, because if that happens, we will be able to create sets of pieces on the board (in particular two sets of pieces, one for black and one for white). This will enable us to hold player states as well (what pieces they have and where). Since `Board.move` swaps `Piece` objects on the board and does not destroy/create them, that makes sense.

I want to try one more time via the `dataclass` decorator by setting both `eq=False` and `unsafe_hast=True` before we move to manual hashing.

**Comment by TrentHawkins** [comment]
I will devise a custom chess position with various pieces all over the board, checking for several movement arrangements and conditions, for testing.

File: src/board.py [hunk 7]
```diff
@@ -154,16 +154,21 @@ def capturable(source_piece: Piece, target: Square):
 
         return _moves
 
-    def move(self, source: Square | str, target: Square | str):
+    def move(self, source_piece: Piece | None, target: Square | str):
```
Comment (prototact) [inline src/board.py]: This is what I was referring to, why call piece when it is `None`. Check in the calling code once instead of every function that takes as argument a piece.

Comment (prototact) [reply src/board.py]: To make it clearer:
```python
add(x: int | None, y: int | None) -> int | None:
    if x is None or y is None
        return None
    return x + y
```
Would you do the above?

**Comment by TrentHawkins** [comment]
[This is the position](https://lichess.org/editor/r1bq1rk1/pp1nbppp/2p2n2/6B1/3QN3/5N2/PPP2PPP/2KR1B1R_w_-_-_5_6?color=white) I used for testing movement constraints. It have various checks. To make this right we should probably make separate unit tests for each piece in specialized conditions that test said piece's specific moves.

File: src/board.py (hunk 7 above)
Comment (TrentHawkins) [reply src/board.py]: This is a sound logic. To tell the truth, most None checks are there because Pylance is whining. I could add a type-ignore instead too, I suppose. I'll follow-up on your suggestion here and let you know ho many checks I trim.

**Comment by TrentHawkins** [comment]
I have reached the end of my abilities with the current code-base, to finish the moves I need to go unto implementing player and game logic (classes `Player` and `Chess`) respectively. This is because players will hold a collection of the pieces they have and the pieces they have captured, so perusing the available moves per player will be far easier on the board. The `color` attribute of `Piece` will become cosmetic. The alternative would be to keep tracking `color` everywhere, but I'd rather implement player logic, for the advanced features I may implement.

Do we merge this work, or do I go on and implement the whole program in one branch? Please review and merge, otherwise tell me what's wrong with it, fix it and merge it, so I can proceed. I may just branch out to proceed, but if your fixes are extensive then I will have to re-adjust all my future work, so I prefer to resolve this now.

File: src/board.py (hunk 7 above)
Comment (TrentHawkins) [reply src/board.py]: This remains to be implemented. As far as `Piece` code and type checking goes, the issue stems from the fact that we chose empty squares to be `None`. `Square` has a similar issue because I have the `Piece.square` attribute, which can be `None` to indicate a that a piece is “captured” or “orphan” (if created outside a board to begin with).

File: src/board.py [hunk 8]
```diff
@@ -127,16 +148,49 @@ def __contains__(self, piece: Piece | None) -> bool:
...
+            def capturable(source_piece: Piece, target: Square):
+                Piece.capturable.__doc__
+                target_piece = self[target]
+
+                if target_piece is not None:
+                    return Piece.capturable(source_piece, target) and source_piece.orientation != target_piece.orientation
+
+                return False
+
+            piece.deployable = MethodType(deployable, piece)
```
Comment (prototact) [inline src/board.py]: I understand this is a way to add context to the `deployable` and `capturable` methods of `Piece`. However, you are still assuming bad input to the methods. If you reject bad input when players are choosing moves, then there is no need for these checks. 
Also, `None` is a general `NoneType` that can have any meaning. I would suggest using a subtype of `Piece`, called `Empty` and make checks using that. This would make code more readable.

File: src/pieces/meleed.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    For a meeled piece (pawns are special) all squares are target squares.
+    """
+
+    @property
+    def moves(self) -> set[Square]:
+        f"""{super().moves.__doc__}"""
+        squares = super().moves
+
+        if self.square is not None:  # If meleed piece is on a board,
```
Comment (prototact) [inline src/pieces/meleed.py]: Everything becomes complicated because you are trying to deal with this issue in the wrong place.

File: src/pieces/special.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+                if self.capturable(square):  # If said target is inside board limits,
+                    squares.add(square)  # Add said target to pawn.
+
+            square = self.square + self.step * self.orientation  # Get forward square,
+
+            if self.deployable(square):  # If said square is inside board limits,
+                squares.add(square)  # Add said square to possible moves,
+
+                if not self.has_moved:  # If the pawn is in its starting position,
```
Comment (prototact) [inline src/pieces/special.py]: This is a good use of Vassilis advice. Although, I would suggest using immutability and replacing the Pawn on the board with a new Pawn for which `has_moved=True` instead of mutating it.

File: src/pieces/special.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    def deployable(self) -> bool:
+        """Check if castling with the two pieces is still possible.
+
+        This check will be refined with context from the board.
+        For now check if either piece has moved, king is unchecked and corresponding rook is still alive.
+
+        Returns:
+            Whether castling with the two pieces is still possible.
+        """
+        return not (self.king.has_moved or self.rook.has_moved or self.king.in_check or self.rook.square is not None)
```
Comment (prototact) [inline src/pieces/special.py]: This is interesting, but I thought deployable was used for another purpose (whether the square was within the board). Of course, the King will always be on the board during the game, so this checks out.

Here you make good use of an invariant, that the king is always on board and you do not test it. 
Why not do the same thing with the rest of the pieces ?

File: src/board.py [hunk 12]
```diff
@@ -21,49 +26,43 @@ class Board:
...
+        if not empty:
+            for color in ["white", "black"]:
+                pawn_rank = (Orientation[color] * 5 + 9) // 2  # Pawn rank per color.
+                main_rank = (Orientation[color] * 7 + 9) // 2  # Main rank per color.
+
+            #   The pawns:
+                for file in Square.file_range:
+                    self[f"{file}{pawn_rank}"] = Pawn(color)
+
+            #   The rest of the pieces:
```
Comment (prototact) [inline src/board.py]: you can condense this by
```python
pieces = [('a', Rook), ('b', Knight), ('c', Bishop,) ('d', Queen), ('e', King), ('f', Bishop), ('g', Knight), ('h', Rook)]
for file, piece in pieces:
    self[f"{file}{main_rank}"] = piece(color)
```

File: src/piece.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+    def __repr__(self) -> str:
+        f"""Represent a {self.__class__.__name__}.
+
+        Returns:
+            The representation of a {self.__class__.__name__}.
+        """
+        return " "  # An unspecified piece is a ghost piece.
+
+    def has_friend(self, other):
```
Comment (prototact) [inline src/piece.py]: makes more sense to call it `is_friend`

File: src/piece.py [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+        """Check weather `other` is a friend of `self`.
+
+        Returns:
+            True if `self` and `other` are of the same color.
+        """
+        return self.orientation == other.orientation if self and other else False
+
+    def has_foe(self, other):
```
Comment (prototact) [inline src/piece.py]: makes more sense to call it `is_foe`

**Review by prototact — APPROVED** [review]
If you want to keep using the `is None` pattern, just ignore my comments. The rest is good, and the idea of checking attributes on the pieces that can change during the game is helpful and will simplify decision-making on the higher layers.

Castling as a separate attribute of the board is also an interesting idea.

File: src/pieces/special.py (hunk 10 above)
Comment (TrentHawkins) [reply src/pieces/special.py]: That's because this flag will only be switched once right? Agreed! I actually have an even better idea. Create a special Pawn class for the starter pawns, which will be replaced by normal pawns after they have moved. This might simplify en-passant logic too. Good remark!

File: src/pieces/special.py (hunk 11 above)
Comment (TrentHawkins) [reply src/pieces/special.py]: I'll look into that now.

File: src/board.py (hunk 8 above)
Comment (TrentHawkins) [reply src/board.py]: Some of these checks are simply because Pylance complains. With the use of invariants like you say in another remark, I can replace several of these checks with the `# type: ignore` Pylance flag.

I really thought of the `Empty` piece at some point, and maybe that of the empty square even, but I would like to avoid it for the moment. Lets see how clean it becomes with the use of `None` for empty, and work from there.

File: src/board.py (hunk 12 above)
Comment (TrentHawkins) [reply src/board.py]: I like this. I will use a `dict` instead, but the idea is the same.

**Comment by TrentHawkins** [comment]
Sorry, I could not wait. I proceeded in implementing player logic. However, this will be used from fundamental threat-detection required for king and king-influenced moves (like pins or defenders or castles), so such changes are still on topic. Have a look. I have not touched the `None` sanity checking yet, I would like to see this together live in a zoom session if possible @prototact. Also, lemme know how you like the game logic as it starts to manifest via the `Player` and `Chess` classes.

**Comment by TrentHawkins** [comment]
I am getting somewhere. I need to redefine `deployable` and `capturable` on two layers. The biggest challenge is implementing pins on the King. What are they? Moves that are nominally legal, but are eventually not, because if performed, they expose the king to a (discovered) check. Which means said pieces are pinned in all ways than the ones that still protect it. For example, if a white bishop defends his King from a black bishop (of the same color of course), the white bishop cannot move away, but it can still move in the direction of the attacking black bishop, so it is partially pinned.

This one will be the trickiest to implement of all, because it needs to check the situation after the desired move is made, so to make the checks, all legit moves must be performed in vitro! It is not too much of work, and I logically see no other way around it. I am close to finishing the implementation!

**Comment by TrentHawkins** [comment]
I did it! Full king safety is implemented! Not in the most efficient way, but in the most clean way. Code is largely self-documenting, and there are comments for the rest. Chess is indeed a very, very complex game. No wonder it stretches the mind so much. We should play more often!

All that remains is to finish implementing castling and en-passant.

Also in other news, I zapped a critter that was hiding in `Board.__setitem__` for a very, very long time! Gotta love the debugger(s)!

**Comment by TrentHawkins** [comment]
Also, a favor: If you can devise more challenging tests for said constraints, please do. Thanks!

File: src/piece.py (hunk 14 above)
Comment (TrentHawkins) [reply src/piece.py]: Removed as unused.

File: src/pieces/special.py (hunk 11 above)
Comment (TrentHawkins) [reply src/pieces/special.py]: This has been adjusted and works now. Please make new comments about it in the latest commit.

File: src/pieces/meleed.py (hunk 9 above)
Comment (TrentHawkins) [reply src/pieces/meleed.py]: I started adding type ignores when necessary, I have to run through the whole code to eliminate any redundant `None` checks. Mind however that not all such checks are unnecessary, I believe about half of them not to be. The particular one you point out here however is.

File: src/board.py (hunk 8 above)
Comment (TrentHawkins) [reply src/board.py]: I did not go through with the empty implementation.

**Comment by TrentHawkins** [comment]
Apart from en-passant (not implemented at all yet) the code is in a very good state right now. If you wait one more day, you may review the whole set of constraints and make final remarks on the code. It will be fixing/improving mode only after I implement en-passant.

File: src/board.py [hunk 15]
```diff
@@ -195,6 +195,38 @@ def __contains__(self, piece: Piece | None) -> bool:
         """
         return any(piece in rank for rank in self.pieces)
 
+    def simulate(self, source_piece: Piece, target: Square | str) -> Generator[Piece | None, None, None]:
```
Comment (prototact) [inline src/board.py]: peek_move is another way to phrase it.

**Comment by TrentHawkins** [comment]
The en-passant rule is turn-sensitive so I had to complete some running parts of the game (for example turn cycling) and a bad, but working move-reader. In theory we can play games now. Work is needed on those rules, before merging this commit.

For now, en-passant successfully detects the ghost squares behind jumping pawns, but I have not implement getting rid of said ghosts if any, and on top (the hardest part) getting rid of the actual enemy pawn, if player takes his en-passant chance. In some sence ghost piece and respective pawn must be linked. Perhaps an internal reference? Not sure that this straightforward, but I have an idea that might work. More on this the coming days.

**Comment by TrentHawkins** [comment]
All done and tested.
